src/util.rs

```diff
@@ -0,0 +1,83 @@
+//! Various utility structures.
+
+use crate::sync::atomic::{AtomicPtr, Ordering};
+use crate::sync::{Arc, AtomicWithMut};
+
+use core::mem::ManuallyDrop;
+use core::ptr;
+
+/// A wrapper around an `Arc` that is initialized atomically in a racy manner.
+#[derive(Default)]
+pub(crate) struct RacyArc<T> {
```

Comment:
How big is the difference between this and `OnceCell` impls?

Reply:
Two main differences:

1). While the once cells are synchronized, this one is racy. While the once cells only run one closure at a time, this one runs all closures at once and sets the value to whichever one comes first.

2). It's only one system word in length, while once cells are usually at least three.

In addition, since our once cells are backed by this crate, we can't use once cells here.

Reply:
We're trying to avoid the `once_cell` crate due to its recent MSRV bump, see smol-rs/async-io#93

Reply:
I'd like to ask about the performance of this, because in the past I implemented https://github.com/zseri/revenq,which I originally based upon a similar pattern and hit a huge perf bottleneck. Thus I ask: How much does this get "hammered" during its lifetime or the lifetime of the chain?

Reply:
After initialization (which is a one-time process), loads from the `AtomicPtr` are a simple atomic load. This is opposed to a `OnceCell`, which consists of a load as well. I doubt that there's much overhead aside from at initialization time. At worst, this leads to the initialization code being run once per thread.

I could replace it with the proper cell to see what happens, but I doubt it would make a difference. Besides, this is what this code used prior to this change.